### ISSUE TITLE
[BUG] Optional lists are marked as unique parameters

### ISSUE THREAD

**Issue by woutermont** [LinkedSoftwareDependencies/Components-Generator.js#74]
Constructor parameters typed as optional list (i.e. the union type of an array type and `undefined`) are converted by the generator to component parameters with `unique: true`, leading Components.js to only load the first element of the list as the parameter itself. 

Removing the `unique: true` statement from the components' JSON-LD description manually solves the problem, but is of course not a long-term solution.

(Edit) Additional diagnosis: this happens from 2.5.0 onwards, and turning back to 2.4.0 or earlier is not possible since back then the generator could not process the union types at all.

**Comment by rubensworks** [comment]
> Constructor parameters typed as optional list (i.e. the union type of an array type and undefined)

So it happens with `param: T[] | undefined`?
Does it also happen with `param?: T[]`?

**Comment by woutermont** [comment]
That is the same, right? We write `param?: T[]`, and typescript interprets that as `param: T[] | undefined`.

**Comment by rubensworks** [comment]
For TS it's indeed the same, but the parser sees it as something different, so the AST won't be the same, which means that the generator may interpret it differently.

**Comment by woutermont** [comment]
What does the parser take as input? Because our `.ts` files (in the case where I discovered it) contain the shorthand version (`param?: T[]`), but TSC already "compiles" that to the longer version (`param: T[] | undefined`) in the `.d.ts` files.

**Comment by rubensworks** [comment]
> What does the parser take as input? 

`.d.ts`

> Because our .ts files (in the case where I discovered it) contain the shorthand version (param?: T[]), but TSC already "compiles" that to the longer version (param: T[] | undefined) in the .d.ts files.

Oh ok, TIL.

Doesn't matter in any case, it's a bug that needs to be fixed :-)

**Comment by rubensworks** [comment]
Found some time to have a look at this, but I can't seem to reproduce the problem myself.

Input:
```typescript
public constructor(param?: string[]) {
  
}
```

Output:
```json
      "parameters": [
        {
          "@id": "custom:lib/Class#Class_param",
          "range": "xsd:string"
        }
      ]
```

Could you share the full constructor that causes the problem? (and any interfaces that it may refer to)

**Comment by woutermont** [comment]
Weird. The full class where I encountered the problem is rather large, but I managed to recreate it with the following.

The class

```
export class TestClass {
  constructor(public testParam?: string[]) {}
}
```

is converted by `tsc` to the type declaration

```
export declare class TestClass {
    testParam?: string[] | undefined;
    constructor(testParam?: string[] | undefined);
}
```

and results in this generated component:

```
{
  "@id": "npmd:testPackage/dist/index#TestClass",
  "@type": "Class",
  "requireElement": "TestClass",
  "extends": [],
  "parameters": [
    {
      "@id": "npmd:testPackage/dist/index#TestClass_testParam",
      "unique": true
    }
  ],
  "constructorArguments": [
    {
      "@id": "npmd:testPackage/dist/index#TestClass_testParam"
    }
  ]
}
```

I'm using v2.6.0 of the generator, with TypeScript v4.4.3.

**Comment by rubensworks** [comment]
The only difference I see is the `public` keyword there, so that may be related to the cause.

**Comment by woutermont** [comment]
True, that appears to cause it, if I add/remove the keyword. Am I correct to expect the same behavior with or without keywords?

**Comment by rubensworks** [comment]
Fixed in 2.6.1. It fixes both the `unique` issue, and the missing `range`.

Note to self: keep in mind that typings in `.d.ts` files may not be identical to those in `.ts`.
